### ISSUE TITLE
Failed to get object masks

### ISSUE THREAD

**Issue by xiaoxiangyeyuwangye** [unrealcv/unrealcv#95]
Hi, guys.
I am trying to use this useful tool to generate some synthetic images，but I had some troubles when I tried to get the object masks.
I downloaded the RealisticRendering pack and type in a command line:
vset /viewmode object_mask
It just became darker，but no object mask displayed. Like this:
![ha9t rolo4 l0jel15 x1iv](https://user-images.githubusercontent.com/32291247/31572847-b4f307ca-b0e1-11e7-8ce8-f3b4cedd2670.png)
But strangely, the other three types of parameter (lit, normal, depth) are valid.
Then I used the jupyter to run the python scripts, still could not obtain the correct object masks.
![yjb_47zxc6m 2ipu r5 kv](https://user-images.githubusercontent.com/32291247/31572877-68f3a824-b0e2-11e7-918d-abfce3705cd4.png)
Is there anything wrong in my operation?
There is still a another problem in "print statistics of this virtual scene and this image". I can't read all the categories in file "object_category.json", only a few categories are loaded. I don't know why.
![2gr04 6 pw80y ni h423](https://user-images.githubusercontent.com/32291247/31572924-3c147d46-b0e3-11e7-805a-4f12d1388602.png)
Thanks for any helps and advises.

**Comment by zfw1226** [comment]
Which platform do you use, Mac, Linux or Windows? 
And Where did you load the RR binary?

**Comment by xiaoxiangyeyuwangye** [comment]
My platform is window 10 64bit，I downlloaded the RR binary files from here
http://docs.unrealcv.org/en/master/reference/model_zoo.html#rr

**Comment by zfw1226** [comment]
Your operation seems right. 
I am sorry for that the problem you met is caused by us when compiling the binary. We will fix it and update new binary. The Linux version RR binary is OK. If you have Device with Linux platform, you can try and enjoy it. 
Thank you for your report.

**Comment by xiaoxiangyeyuwangye** [comment]
OK！Thanks for your reply. This is really a useful project and a creative idea. I'm looking forward it getting more exciting. I will keep attention on your guys' wonderful work, and hope I can help it getting better.

**Comment by xiaoxiangyeyuwangye** [comment]
Is the second problem caused by the first one? I can't read all the categories in file "object_category.json".

**Comment by xiaoxiangyeyuwangye** [comment]
In my opinion, I don't think think there is any relationship between these two problems.

**Comment by zfw1226** [comment]
Yeah, the second problem seems caused by the first one. 
We have updated the RealisticRendering-Windows binary. You can download it from the model zoo and try it again. 
Have a good time.

**Comment by xiaoxiangyeyuwangye** [comment]
Yeah, it works correctly. Could you please tell me where should if I get start I want to make something like unrealcv，or add some other functions upon unrealcv. I feel really confused about the rendering relevant c++ API and its corresponding documents on their web. Is there any helpful tutorial or document on this issue? Anyway, Thanks a lot for your reply and help!

**Comment by kunaltyagi** [comment]
Issue exists in Unreal Engine 4.18, Linux.

Built Engine as well as unrealcv from source (unrealcv is in both Engine/Plugins/Runtime as well as Project/Plugins). 

Everything gets greyed out on object_mask, but depth, lit and normal mode are displaying correct output. 

![object-mask-fail](https://user-images.githubusercontent.com/2657068/33540758-ce11d6ae-d90f-11e7-8210-225239e93db6.png)

Also, could we have a documented C++ API/tutorial? I'd like to send current state of actors along with the synced image. IMO this should be possible by overloading the server and adding more commands. But how?

Is it possible to send more than one (depth + lit) image from same scene? My scene isn't static... :cry: 

Plus, kudos to the idea and dev team. :+1:

**Comment by deitylink1** [comment]
Similar problem, UnrealCV running the "playground" -  the object mask just grays everything in an unlit fashion. (also depth works, but makes the sky very weird as well. Not that that matters too much for me)

![image](https://user-images.githubusercontent.com/32622655/33860546-65b8c6ba-dea7-11e7-83b8-2376faf0078d.png)

![image](https://user-images.githubusercontent.com/32622655/33860579-8a0d8c8a-dea7-11e7-8895-37fd50158cce.png)

I updated the project to 4.18 myself, but think I noticed something on git last night about a version where that is already done. Perhaps this was already fixed there?

**Comment by Jyouhou** [comment]
Hi @deitylink1 ,

I encountered the same issue. When I sent 'vset /viewmode object_mask', the scene just turned gray. 'vget /camera/0/object_mask xxx.png' didn't write anything to my disk as well.

I am wondering if you have solved this problem or not. :)

**Comment by qiuwch** [comment]
Hi, @Jyouhou 

What is your UE4 version?

**Comment by Jyouhou** [comment]
> Hi, @Jyouhou
> 
> What is your UE4 version?

Hi @qiuwch ,

I use UE4.18 on Ubuntu 16.04. The version of my UnrealCV is 0.3.10.

**Comment by qiuwch** [comment]
If you use vget /camera/0/object_mask png. Will it give you a binary response?

The png file location is relative to the binary file, so it might be deeply buried in a folder.

Does image capture work? I mean vget /camera/0/lit filename.

**Comment by Jyouhou** [comment]
Hi @qiuwch ,

Yes, it does return a binary response. But the response is black&white. So what matters most is that the returned object mask does not contain meaningful information... 

Since I use absolute path only, I believe it's not because the file has been put somewhere else. Besides, the same set of commands (lit/depth/normal) work perfect under the same setting and same filepath.

**Comment by qiuwch** [comment]
I think I remember the reason.

It is because of a breaking change of UE4 when the engine developer updates the engine from 4.16 to 4.17.

The solution are:
1. Use 4.16 and master branch of unrealcv
2. Use > 4.16 version (recommend 4.19) and use fast branch of unrealcv.

I highly recommend using 4.19 and the fast branch, which is what we are using internally in our lab.

The details of the breaking change is as follows, which you don't need to understand and I wrote here for a memo. Unrealcv previously uses VertexColor paint to generate segmentation mask and this feature has been removed from the engine level. So I wrote a completely different way to generate ground truth which uses "AnnotationComponent.cpp" to generate segmentation mask. If you can find "Source/UnrealCV/Private/Component/AnnotationComponent.cpp", then that is the correct branch.

**Comment by Jyouhou** [comment]
Hi @qiuwch

**Comment by Jyouhou** [comment]
Hi @qiuwch 

Thanks for your reply! 

Yep, the slow speed is another issue I am faced with. Earlier I noticed the "fast" branch from this thread 
[#109](https://github.com/unrealcv/unrealcv/issues/109#issuecomment-358886970).

I guess this fast version can help. 

Thanks!

**Comment by Jyouhou** [comment]
Hi @qiuwch ,

I am experimenting with 4.18. In case other people encounter similar issues, I would like to write it down here.

The current fast branch (7d68129) relies on a header called "Runtime/Engine/Public/Rendering/SkeletalMeshRenderData.h", which does not exist in version 4.18. Therefore, it would fail to compile for UE4.18. 4.19 should solve the problem

**Comment by Jyouhou** [comment]
Hi @qiuwch ,

I am experimenting with 4.18. In case other people encounter similar issues, I would like to write it down here.

The current fast branch (7d68129) relies on a header called "Runtime/Engine/Public/Rendering/SkeletalMeshRenderData.h", which does not exist in version 4.18. Therefore, it would fail to compile for UE4.18. 4.19 should solve the problem.

It would be greatly appreciated if a reminder about this is added to the documentation.
